addons/mail/models/res_users.py

```diff
@@ -268,42 +268,45 @@ def _init_messaging(self):
     @api.model
     def systray_get_activities(self):
         activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
-        activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
-        for activity in activities:
-            record = self.env[activity.res_model].browse(activity.res_id)
-            activities_by_record_by_model_name[activity.res_model][record] += activity
-        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_model_name.keys()})
-        user_activities = {}
-        for model_name, activities_by_record in activities_by_record_by_model_name.items():
-            domain = [("id", "in", list({r.id for r in activities_by_record.keys()}))]
-            allowed_records = self.env[model_name].search(domain)
-            if not allowed_records:
-                continue
-            module = self.env[model_name]._original_module
-            icon = module and modules.module.get_module_icon(module)
-            model = self.env["ir.model"]._get(model_name).with_prefetch(model_ids)
-            user_activities[model_name] = {
-                "id": model.id,
-                "name": model.name,
-                "model": model_name,
-                "type": "activity",
-                "icon": icon,
-                "total_count": 0,
-                "today_count": 0,
-                "overdue_count": 0,
-                "planned_count": 0,
-                "actions": [
-                    {
-                        "icon": "fa-clock-o",
-                        "name": "Summary",
-                    }
-                ],
-            }
-            for record, activities in activities_by_record.items():
-                if record not in allowed_records:
+        activities_by_record_by_module_by_model_name = {}
+        group_custom_info = {}
+        models = list(dict.fromkeys(activities.mapped('res_model')))
+        for model in models:
+            activities_by_record_by_module_by_model_name[model], group_custom_info[model] = self.env[model]._sort_activities_by_record_by_module(
+                activities.filtered_domain([('res_model', '=', model)]))
+        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_module_by_model_name})
+        user_activities = []
+        for model_name, activities_by_record_by_module in activities_by_record_by_module_by_model_name.items():
+            for module, activities_by_record in activities_by_record_by_module.items():
+                domain = [("id", "in", list({r.id for r in activities_by_record.keys()}))]
+                allowed_records = self.env[model_name].search(domain)
```

Comment:
To avoid the search in the 2nd for loop we could maybe do that:
```suggestion
            all_record_ids = set()
            record_ids_per_module = {
                module
                
                for module, activities_by_record in activities_by_record_by_module.items()
            }
            for module, activities_by_record in activities_by_record_by_module.items():
                record_ids = list({
                    r.id 
                    for r in activities_by_record.keys()
                })
                all_record_ids.union(record_ids)
                record_ids_per_module[module] = record_ids
            domain = [("id", "in", list(all_record_ids))]
            all_allowed_records = self.env[model_name].search(domain)
            for module, activities_by_record in activities_by_record_by_module.items():
                record_ids = record_ids_per_module[module]
                allowed_records = all_allowed_records.filtered(lambda r: r.id in record_ids)
```

We can obviously update my suggestion if the key in the dict will be the id of the record as I suggest in my previous comment.

I don't know if it is overkill or not, probably yes